### DISCUSSION TITLE
Need to hook into CodeSpaces Delete Lifecycle Event

### DISCUSSION THREAD

**Discussion by r3dpointsp** [community/community#11578] [Codespaces]
We provision a 3rd party SaaS Client ID during `onCreate` which works great and really reduces developer friction on getting all the 3rd party keys and integrations set up before development. Now we have a new problem - we are creating many Client IDs and now getting automated emails from our 3rd party vendors that we have unresponsive Client Apps - since those apps are being deleted or timed out by CodeSpaces.

Can we get access to the onDelete lifecycle event in devcontainer.json? This would allow us to call the 3rd party SaaS and de-provision the current Client ID and clean up any resources we need cleaning up.

Also as a Repository and Organization owner it would be great to query across all CodeSpaces / get webhooks for all CodeSpaces lifecycle events so we can run GitHub actions on those events. Also knowing a specific CodeSpace had a lifecycle event isn't as useful as being able to operate on that CodeSpace even though I am not that CodeSpace owner - especially in the onDelete case.

Thanks!

**Comment by r3dpointsp** [comment]
Bump since no response in 28 days. Thanks!

**Comment by shrink** [comment]
I'm just a bystander, not from GitHub, but I think you have a challenge here that won't be resolved by the `onDelete` lifecycle event. A Codespace can exist indefinitely without activity, meaning the third-party inactivity check will fail _even when a Codespace continues to exist_. The problem of third-party Codespace dependencies is, I think, inextricably linked to the way that Codespaces work.

You may wish to consider whether an alternative is to change how you provision these resources: instead of provisioning on Codespace creation (which doesn't relate to all activity) provision on Codespace start (which does relate to all activity) and then monitor the Codespace for activity yourself to determine when to de-provision these resources (i.e: Codespace is now inactive -> de-provision).

Although a little more cumbersome, it would provide a reliable solution to your problem.

**Comment by rayjlinden** [comment]
I'd really like this feature as well.  While shrink's comment is valid.  We would need to think about cleaning orphaned code spaces as well but that is a different issue...  But being able to cleanup allocated resources associated with a code space (but not in the code space) would be very useful!

In my case, it is some AWS resources I allocate and want to clean up with each codespace.
